shpc/main/container/base.py

```diff
@@ -175,5 +178,28 @@ def get_features(self, config_features, settings_features, extra=None):
 
         return features
 
+    def _generate_wrapper_scripts(self, wrapper_template, aliases, module_dir, features, command=None, container_sif=None, image=None, tty=None):
```

Comment:
@vsoch , so here it is, the new function `_generate_wrapper_scripts` to modularise the code that generates wrapper scripts. It gets called by the singularity- and docker- specific `install` methods.

Please have a look at this bit of code (especially the interface) and the corresponding method calls, to be honest I don't write OO code in Python often, so I don't have a good eye for best practices, nor conventions specific to this project :)

Reply:
Let me think about this - we are hitting a pattern of "here is a template and I want to render all the things into it" and my worry is that if different wrapper scripts are added, we aren't making it easy to do that, optionally supplying custom variables to one but not the other. I do think we want a different class to describe a wrapper script, and then the main install functions will just do:

```python
from shpc.main.wrappers import generate_wrappers

if self.settings.wrapper_scripts:
    generate_wrappers(...)
``` 
And we'd provide all the general settings / config / container / paths to this function, and each wrapper template could deal with them as follows (e.g., for this first example you are requiring aliases but I don't see why all wrappers need that).

The design also needs to take into account that an install could have more than one wrapper or we might want to add a wrapper to an existing install:

```bash
$ shpc wrapper add <name> <module>
```
And then there would be a wrapper base class with a standard set of functions (that can be subclassed if needed) and we would have a master lookup.

So - this is a lot of work I suspect for a PR to ask, and I'm not even sure it will work, so if it's okay with you I can pull your branch and take a first shot.

Reply:
What I'm trying to wrap my head around now is how to generalize, given that singularity.sh and docker.sh seem to be tied up tightly with the lua/tcl templates.

Reply:
Adding some of my thoughts, too, maybe they can be good input in the design of the new class.

The new wrapper scripts are an alternative way to the shell aliases/functions, for exposing container functionalities to users.

Some of the key differences I am aware of:
* wrapper scripts: can be called by other processes, in particular `mpirun` (mandatory for MPI applications), and scheduler managers such as Slurm `srun` (again mandatory for MPI, but also recommended for any application in a Slurm cluster); shell aliases/functions will throw an error if used this way
* shell aliases/functions: self-contained in the module, therefore allowing for a neater, more concise, more manageable deployment of the user interface to the containers; wrapper scripts on the other hand are files sitting on the filesystem (not a huge  deal, but still something to be aware of)

On the other hand, in terms of customisability, I think wrappers and shell aliases are quite similar; we realised the opportunity of customisability when working on the former, but I think that almost any edit that is allowed by the wrappers can also be implemented in the form of aliases.

In this regard, having more flexibility with both wrapper templates and tcl/lua module templates matches the customisability that is inherent in both.
Of course, module templates have even more room for customisation, as they contain more than just the command aliases.

So, that's probably why in my mind the two are quite tied up. That's also why I see the benefit of having a dedicated class, as there's quite some generality that can be expressed here.

However, I am also wondering whether there can be other types of templates or wrappers, that may expand this picture even further...? 
One thing that came to my mind several times are container-specific additional wrapper templates, that apply to specific commands. For instance, in a Rstudio container, a specific wrapper for rstudio-server (or whatever it's called), to fully customise the spawning of an Rstudio server instance (or same for Jupyter hub). 
In other words, a registry directory for a container might end up having one or all of the following files: container.yaml (mandatory), docker.sh (custom docker template for any command), singularity.sh (custom singularity template for any command), <command>-docker.sh (template for a specific <command>), <command>-singularity.sh.
Note I have been thinking out loud in the last paragraph, not sure yet whether the corner case is worth any extra work in the code.

That's all for now!

Reply:
Yes, I'm still awake! Being very bad staying up :) I have most of the work done and will PR shortly.